sql: Timeout in TestTxnUserRestart

https://circleci.com/gh/cockroachdb/cockroach/16732

```
goroutine 10198 [IO wait]:
net.runtime_pollWait(0x7fe83e9f7540, 0x72, 0xc820eb4c00)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc8212852c0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8212852c0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc821285260, 0xc820eb4c00, 0x400, 0x400, 0x0, 0x7fe83e9dd050, 0xc820010048)
	/usr/local/go/src/net/fd_unix.go:250 +0x23a
net.(*conn).Read(0xc820026170, 0xc820eb4c00, 0x400, 0x400, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
crypto/tls.(*block).readFromUntil(0xc82129ea80, 0x7fe83e9f99c0, 0xc820026170, 0x5, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:460 +0xcc
crypto/tls.(*Conn).readRecord(0xc8206bd200, 0x1b76217, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:562 +0x2d1
crypto/tls.(*Conn).Read(0xc8206bd200, 0xc8207ef000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/tls/conn.go:939 +0x167
bufio.(*Reader).fill(0xc820d50240)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Read(0xc820d50240, 0xc8201da020, 0x5, 0x200, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:207 +0x260
io.ReadAtLeast(0x7fe83e9ea070, 0xc820d50240, 0xc8201da020, 0x5, 0x200, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:297 +0xe6
io.ReadFull(0x7fe83e9ea070, 0xc820d50240, 0xc8201da020, 0x5, 0x200, 0xc820a9e140, 0x0, 0x0)
	/usr/local/go/src/io/io.go:315 +0x62
github.com/cockroachdb/pq.(*conn).recvMessage(0xc8201da000, 0xc820a9e140, 0x5aa8b2, 0x0, 0x0)
	/go/src/github.com/cockroachdb/pq/conn.go:953 +0x15a
github.com/cockroachdb/pq.(*conn).recv1Buf(0xc8201da000, 0xc820a9e140, 0x52)
	/go/src/github.com/cockroachdb/pq/conn.go:1003 +0x2f
github.com/cockroachdb/pq.(*conn).recv1(0xc8201da000, 0xc8207b16c0, 0xc820a9e140)
	/go/src/github.com/cockroachdb/pq/conn.go:1024 +0x9b
github.com/cockroachdb/pq.(*conn).simpleExec(0xc8201da000, 0xc820416000, 0x4d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/pq/conn.go:614 +0x2a1
github.com/cockroachdb/pq.(*conn).Exec(0xc8201da000, 0xc820416000, 0x4d, 0x25cc7c8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/pq/conn.go:864 +0x135
database/sql.(*DB).exec(0xc8208b0e70, 0xc820416000, 0x4d, 0x0, 0x0, 0x0, 0x800701, 0x0, 0x0, 0x0, ...)
	/usr/local/go/src/database/sql/sql.go:1035 +0x2c2
database/sql.(*DB).Exec(0xc8208b0e70, 0xc820416000, 0x4d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/database/sql/sql.go:1009 +0xbe
github.com/cockroachdb/cockroach/sql_test.abortTxn(0xc820450000, 0xc8208b0e70, 0x0)
	/go/src/github.com/cockroachdb/cockroach/sql/txn_restart_test.go:391 +0x253
github.com/cockroachdb/cockroach/sql_test.runTestTxn(0xc820450000, 0xc82097ad20, 0x17ee1a0, 0xf, 0xc8205a3d27, 0xc8208b0e70, 0xc820214000, 0xc820a5c428)
	/go/src/github.com/cockroachdb/cockroach/sql/txn_restart_test.go:362 +0x3f0
github.com/cockroachdb/cockroach/sql_test.TestTxnUserRestart.func2(0xc820214000, 0x18fd620)
	/go/src/github.com/cockroachdb/cockroach/sql/txn_restart_test.go:447 +0x5f
github.com/cockroachdb/cockroach/sql_test.exec(0xc820450000, 0xc8208b0e70, 0xc8205a3f18)
	/go/src/github.com/cockroachdb/cockroach/sql/txn_restart_test.go:330 +0x27f
github.com/cockroachdb/cockroach/sql_test.TestTxnUserRestart(0xc820450000)
	/go/src/github.com/cockroachdb/cockroach/sql/txn_restart_test.go:448 +0x88a
testing.tRunner(0xc820450000, 0x22cb348)
	/usr/local/go/src/testing/testing.go:473 +0x98
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0x892
```

Another instance of the TestErrorOnCommit failure:
https://circleci.com/gh/cockroachdb/cockroach/16874

```
=== RUN   TestErrorOnCommit
W160426 14:22:24.409511 gossip/gossip.go:892  not connected to cluster; use --join to specify a connected node
..
W160426 14:22:49.916274 storage/store.go:1715  conflicting intents on [/Table/51/1/0/0 /Table/51/1/0/2/1]: resolved? false
panic: test timed out after 1m10s
```

--- a/sql/txn_restart_test.go
+++ b/sql/txn_restart_test.go
@@ -24,6 +24,7 @@ import (
 	"testing"
 
 	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/server"
 	"github.com/cockroachdb/cockroach/testutils"
 	"github.com/cockroachdb/cockroach/testutils/storageutils"
 	"github.com/cockroachdb/cockroach/util/caller"
@@ -314,6 +315,8 @@ BEGIN;
 
 // exec takes a closure and executes it repeatedly as long as it says it needs
 // to be retried.
+// This function needs to be called from tests that set
+// server.Context.TestingKnobs.ExecutorTestingKnobs.FixTxnPriority = true
 // TODO(andrei): change this to return an error and make runTestTxn inspect the
 // error to see if it's retriable once we get a retriable error code.
 func exec(t *testing.T, sqlDB *sql.DB, fn func(*sql.Tx) bool) {
@@ -321,7 +324,6 @@ func exec(t *testing.T, sqlDB *sql.DB, fn func(*sql.Tx) bool) {
 	if err != nil {
 		t.Fatal(err)
 	}
-	// TODO(andrei): make the priority deterministic here once Radu introduces that syntax.
 	if _, err := tx.Exec(
 		"SAVEPOINT cockroach_restart; SET TRANSACTION PRIORITY LOW;"); err != nil {
 		t.Fatal(err)
@@ -379,16 +381,22 @@ func runTestTxn(t *testing.T, magicVals *filterVals, expectedErr string,
 // that had an intent on that key.
 // This cannot be done as an injected error, since we want the pusher to clean
 // up the intents of the pushee.
+// This function needs to be called from tests that set
+// server.Context.TestingKnobs.ExecutorTestingKnobs.FixTxnPriority = true
 func abortTxn(t *testing.T, sqlDB *sql.DB, key int) {
-	// Execute all statements in one batch so that the server retries it
-	// automatically in case our txn (the pusher) fails to push the victim. Even
-	// though it starts with a HIGH priority, that's no guarantee it'll always
-	// succeed from the first try.
-	// TODO(andrei): make the priority deterministic here once Radu introduces
-	// that syntax and then we won't need all the statements in one batch.
-	deleteCmd := fmt.Sprintf("DELETE FROM t.test WHERE k = %d", key)
-	if _, err := sqlDB.Exec(
-		fmt.Sprintf("BEGIN; SET TRANSACTION PRIORITY HIGH; %s; COMMIT;", deleteCmd)); err != nil {
+	var err error
+	var tx *sql.Tx
+	tx, err = sqlDB.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := tx.Exec("SET TRANSACTION PRIORITY HIGH"); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := tx.Exec("DELETE FROM t.test WHERE k = $1", key); err != nil {
+		t.Fatal(err)
+	}
+	if err = tx.Commit(); err != nil {
 		t.Fatal(err)
 	}
 }
@@ -400,6 +408,7 @@ func TestTxnUserRestart(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	ctx, cmdFilters := createTestServerContext()
+	ctx.TestingKnobs.ExecutorTestingKnobs.FixTxnPriority = true
 	server, sqlDB, _ := setupWithContext(t, ctx)
 	defer cleanup(server, sqlDB)
 
@@ -515,7 +524,9 @@ CREATE DATABASE t; CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT);
 func TestErrorOnCommit(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
-	server, sqlDB, _ := setup(t)
+	ctx := server.NewTestContext()
+	ctx.TestingKnobs.ExecutorTestingKnobs.FixTxnPriority = true
+	server, sqlDB, _ := setupWithContext(t, ctx)
 	defer cleanup(server, sqlDB)
 	if _, err := sqlDB.Exec(`
 CREATE DATABASE t; CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT);
@@ -527,7 +538,6 @@ CREATE DATABASE t; CREATE TABLE t.test (k INT PRIMARY KEY, v TEXT);
 	if err != nil {
 		t.Fatal(err)
 	}
-	// TODO(andrei): make the priority deterministic here once Radu introduces that syntax.
 	if _, err := tx.Exec("SAVEPOINT cockroach_restart; SET TRANSACTION PRIORITY LOW;"); err != nil {
 		t.Fatal(err)
 	}

